### PR TITLE
Feature/two new interfaces

### PR DESCRIPTION
**Pull request by mwittie**
In this PR I addressed the comments from 
https://github.com/blocky/nitriding/pull/7
https://github.com/blocky/nitriding/pull/8
I'm going to decline those PRs, so we can focus on this one without getting bogged down with the other code. And, I'll let this PR finish, so the scope doesn't creep on it.

Incidentally, we need to add more interfaces than there are in this PR to support the new brave/nitriding functionality the pushed recently. But I'm going to skip those for now, to make the PR work incremental. 

Major changes are:

- I changed how we handle certificate files. There's a separate package for them now. We also don't need to pass file name strings around.
- I added some common testing helpers to `utils`
- I added Fuzz tests for many of the input tests to avoid using random inputs
- I reworked the `Validator` interface to support path validation, which we need for some of the Fuzz tests, since it's possible to create, but not remove files with some randomly generated names. So we check for viable paths before creating files.

### CODE REVIEW

File: certificate/self_sig_cert.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		&pem.Block{Type: "CERTIFICATE", Bytes: cert.derBytes},
+	)
+	if err != nil {
+		return nil, err
+	}
+	err = pemFile.Close()
+	if err != nil {
+		return nil, err
+	}
+	err = os.Chmod(pemFile.Name(), 0o400)
```
Comment (ianhecker) [inline certificate/self_sig_cert.go:165]: you should be able to just put `0400` instead of `0o400`

Comment (ianhecker) [reply certificate/self_sig_cert.go:165]: https://pkg.go.dev/io/fs#FileMode

File: certificate/self_sig_cert_test.go [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+			assert.NoError(t, err)
+			assert.NotNil(t, privateKey)
+			pemBytes, err := cert.ToMemory()
+			assert.NoError(t, err)
+
+			certificate.VerifyCert(t, cert.DerBytes(), pemBytes)
+
+			parsedCert, err := x509.ParseCertificate(cert.DerBytes())
+			assert.NoError(t, err)
+			assert.Equal(t, parsedCert.PublicKey, &privateKey.PublicKey)
```
Comment (ianhecker) [inline certificate/self_sig_cert_test.go:99]: would be helpful human parsing during a testing error to resolve these objects' addresses rather than compare pointers
```golang
assert.Equal(t, *parsedCert.PublicKey, privateKey.PublicKey)
```
then you can see the object rather than the pointer when testing errors out - makes it easier to read

File: certificate/self_sig_cert_test.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+		} else {
+			assert.Equal(t, certificate.SelfSigCert{}, cert)
+			assert.Truef(
```
Comment (ianhecker) [inline certificate/self_sig_cert_test.go:102]: this is a long assertion - can we make a little helper function somewhere?

```golang
func stringInvalidOrUnencodeable(err) error {
    invalid := utils.ErrContStrs(err, "asn1", "string not valid")
    unencodeable := utils.ErrContStrs(err, "x509", "cannot be encoded")
    if invalid || unencodeable {
        return fmt.Errorf("certOrg '%v', fqdn '%#v', error '%v'", certOrg, fqdn, err)
    }
    return nil
}
...
assert.Nil(t, stringInvalidOrUnencodeable(err))
```

File: certificate/self_sig_cert_test.go [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+	})
+}
+
+func TestMakeSelfSignCert_FromFile_HappyPath(t *testing.T) {
+	certFile := utils.ReflectTestName() + ".pem"
+
+	origCert, err := certificate.MakeSelfSigCert("", "")
+	assert.NoError(t, err)
+	assert.NotNil(t, origCert)
+	assert.NotNil(t, origCert.Digest())
```
Comment (ianhecker) [inline certificate/self_sig_cert_test.go:192]: add newline for readability

File: certificate/self_sig_cert_test.go [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+	assert.NotNil(t, privateKey)
+
+	_, err = origCert.ToFile(certFile)
+	assert.NoError(t, err)
+
+	certCopy, err := certificate.MakeSelfSignCertFromFile(certFile)
+	assert.NoError(t, err)
+	assert.NotNil(t, certCopy)
+	assert.Equal(t, origCert.DerBytes(), certCopy.DerBytes())
+	assert.Equal(t, origCert.Digest(), certCopy.Digest())
```
Comment (ianhecker) [inline certificate/self_sig_cert_test.go:204]: add newline

File: utils/utils.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package utils
+
+import (
+	"runtime"
+	"strings"
+)
+
+func ErrContStrs(err error, subStrs ...string) bool {
```
Comment (ianhecker) [inline utils/utils.go:8]: May be more readable if it unabbreviated `ErrorContainsStrings` - i didnt know what the `Cont` meant reading this function in other places

File: utils/utils.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	for _, subStr := range subStrs {
+		if !strings.Contains(err.Error(), subStr) {
```
Comment (ianhecker) [inline utils/utils.go:10]: if you pass in an error that is nil and call `err.Error()` on it, it will cause a panic. you have to check that `err` is non-nil before calling `err.Error()`

Comment (ianhecker) [reply utils/utils.go:10]: the easiest way to avoid this would be https://go.dev/play/p/XqL4NNWY7bS

File: validation/path_validator_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+// NOTE: This test uses global (file) state and cannot run on multiple workers
+// Run it with go test -fuzz=FuzzFileValidator_Validate -parallel=1
+func FuzzFileValidator_Validate(f *testing.F) {
+	tests := []struct {
+		fileName string
+	}{
+		{" "},   // file name not valid
+		{"0/0"}, // no such file or directory
+		{"000000000000000000000000000000000000000000000000000000000000000000" +
```
Comment (ianhecker) [inline validation/path_validator_test.go:19]:
```golang
...
{" "},
{"0/0"},
{strings.Repeat("0", 256)}
...
```

Comment (ianhecker) [reply validation/path_validator_test.go:19]: placing that many zeros as a test name is not very readable, and it loses the important thing - how many zeros is it to be too long? what is the edge case? is 255 just right? and 256 just one too long?

File: validation/path_validator_test.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	for _, tc := range tests {
+		f.Add(tc.fileName)
+	}
+
+	fv := validation.MakeFileValidator(validation.LinuxFilePathRegex)
+
+	f.Fuzz(func(t *testing.T, fileName string) {
+		err := fv.Validate(fileName)
+		if err != nil {
+			assert.Truef(
```
[the remark this answers is not in the document]
Comment (ianhecker) [reply validation/path_validator_test.go:35]: i think that https://stackoverflow.com/questions/60140859/does-go-have-file-name-restrictions provides enough clarity on edge cases that we can write specific tests without having to fuzz

File: validation/path_validator_test.go (hunk 8 above)
Comment (mwittie) [reply validation/path_validator_test.go:19]: Great! Done

File: validation/path_validator_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (mwittie) [reply validation/path_validator_test.go:35]: It's not just the file path string, but also whether the file can be created. For example fuzzing showed me that a file `/a` gets permission denied, while `a/a` gets a no such directory error.

Having said that I'm going to follow what we discussed today in engineering and provide table driven tests for the test cases discovered by fuzzing. That way we test that each case results in an specific expected error, not just some set of errors defined by the `||` statement.

File: utils/utils.go (hunk 6 above)
Comment (mwittie) [reply utils/utils.go:8]: Checking for nil and replaced with `ErrorMatchesPattern`

File: utils/utils.go (hunk 7 above)
Comment (mwittie) [reply utils/utils.go:10]: Checking for nil and replaced with `ErrorMatchesPattern`

File: certificate/self_sig_cert_test.go (hunk 5 above)
Comment (mwittie) [reply certificate/self_sig_cert_test.go:204]: done

File: certificate/self_sig_cert_test.go (hunk 4 above)
Comment (mwittie) [reply certificate/self_sig_cert_test.go:192]: done

File: certificate/self_sig_cert_test.go (hunk 2 above)
Comment (mwittie) [reply certificate/self_sig_cert_test.go:99]: yeah, would be better, but dereferencing `*parsedCert.PublicKey` does not work. You get the error `Invalid indirect of 'parsedCert.PublicKey' (type 'any')`. Under the circumstances I thought comparing through pointers was fine.

File: certificate/self_sig_cert_test.go (hunk 3 above)
Comment (mwittie) [reply certificate/self_sig_cert_test.go:102]: I replaced these assertions with 
```
ok, err := utils.ErrorMatchesPattern(
				makeErr,
				"asn1: string not valid UTF-8",
				"x509: .* cannot be encoded as an IA5String",
			)
assert.NoError(t, err)
assert.Truef(t, ok, "certOrg '%v', fqdn '%#v', error '%v'", certOrg, fqdn, makeErr)
```

I'm not sure if that's shorter, but I think it's more readable and flexible in catching error patterns.

File: certificate/self_sig_cert.go (hunk 1 above)
Comment (mwittie) [reply certificate/self_sig_cert.go:165]: done

File: certificate/self_sig_cert.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	}
+}
+
+func MakeSelfSignCertFromFile(
+	fileName string,
+) (
+	SelfSigCert,
+	error,
+) {
+	bytes, err := os.ReadFile(fileName)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/self_sig_cert.go:101]: That's a great point. There is a limit reader interface that's in the next PR. It's something that came with some of the newer functionality from Brave, so I was waiting to apply introduce it.
But, I forgot/didn't realize we already had this interface in ZPR. So, I'm going to use what we already have!

File: certificate/cert.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+type (
+	Digest   [32]byte
+	DerBytes []byte
+	PemBytes []byte
+)
+
+func pemToDer(pemBytes PemBytes) (DerBytes, error) {
+	rest := []byte{}
+	for rest != nil {
+		block, rest := pem.Decode(pemBytes)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert.go:24]: Well, to answer your question the inside the loop declaration gets a priority. In 
```
	rest := []byte{}
	var block *pem.Block
	for rest != nil {
		block, rest := pem.Decode(pemBytes)
```
The with `block` gets the unused variable error.

I changed the assigned for clarity of scope for both of these variables.

File: certificate/acme_cert.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	return ACMECert{
+		derBytes:       derBytes,
+		digest:         digest,
+		encodeToMemory: encodeToMemory,
+	}
+}
+
+func (cert ACMECert) PrivateKey() (*ecdsa.PrivateKey, error) {
+	return nil,
+		errors.New("cert privateKey is nil for ACMECert")
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/acme_cert.go:34]: The certificates, as will become apparent in the next PR where I will introduce the `attestation` package, need to be interchangeable. But, the ACME will never have a private key.

I really like your suggestion of sub interfaces though, so that's what I'm going to do.

File: certificate/acme_cert.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (cert ACMECert) ToFile(_ string) (*os.File, error) {
+	return nil, errors.New("ACMECert.ToFile is not meant to be called")
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/acme_cert.go:38]: done

File: utils/limit_reader.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
+package utils
+
```
Comment (ianhecker) [inline utils/limit_reader.go:2]: sorry @mwittie i actually forgot that we copied over `LimitedReader` to zipperchain, and i made some improvements!

https://bitbucket.org/blockymcchainerson/zpr/src/main/nub/input/reader.go

Comment (ianhecker) [reply utils/limit_reader.go:2]: i think you'll find some improvements, like the `exceeds-size.json` file is now not over a kilobyte long

File: utils/testdata/exceeds-size.json [hunk 15]
```diff
@@ -0,0 +1,128 @@
+[
```
Comment (ianhecker) [inline utils/testdata/exceeds-size.json:1]: haha using the new `LimitedReader` from zipperchain the maximum size of files becomes injectable - and this file is shrunk down to a single line

File: validation/path_validator_test.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+func TestMakeFileValidator_Validate_Errors(t *testing.T) {
+	tests := map[string]struct {
+		fileName string
+		errRegex string
+	}{
+		"invalid fileName":  {fileName: " ", errRegex: "file name not valid"},
+		"no such file":      {fileName: "0/0", errRegex: "open .* no such file"},
+		"fileName too long": {fileName: strings.Repeat("0", 256), errRegex: "open .* too long"},
+		"permission denied": {fileName: "/0", errRegex: "open .* permission denied"},
```
Comment (ianhecker) [inline validation/path_validator_test.go:28]: very nice test cases, my ADHD thanks you

**Review by ianhecker — APPROVED** [review]

File: utils/limit_reader.go [hunk 17]
```diff
@@ -0,0 +1,68 @@
+package utils
```
[the remark this answers is not in the document]
Comment (dlm) [reply utils/limit_reader.go:1]: I would caution about creating a "utils" package.  Often, utils packages tend to turn random and unfocused.  

Also, I am a little concerned about copying code from one ACTIVE project to ACTIVE active project (key here is active), especially one that we are considering releasing in a PR to another organization. Here are my concerns:

1. (Somewhat obvious), if there is a bug in the copied code we have to track it down and fix it in multiple places and projects.
2. (Somewhat more subtle)  Depending on the license of the project we are copying it into, there could be some ambiguity about copyright infringement.  Say for example, we copy limited reader from zpr here.  In 6 months, we make another project (lets call it projectX) that needs limited reader, copy limited reader from zpr to projectX, is that okay?  I don't know, but I am not sure the effort of finding out is worth it.
 
If we are finding that we are building tools that are useful in many of our projects, we should probably package those into libraries.  If those libraries do not contain any BLOCKY secret sauce, we can release them publicly.

File: utils/utils.go [hunk 18]
```diff
@@ -0,0 +1,33 @@
+package utils
+
+import (
+	"errors"
+	"regexp"
+	"runtime"
+	"strings"
+)
+
+func ErrorMatchesPattern(err error, regexStrs ...string) (bool, error) {
```
Comment (dlm) [inline utils/utils.go:10]: Because of the file structure of go, it is hard to create test helper functions used in multiple locations of the library. To address, the convention is to put the code in a "test" package that is not imported by anything other than test code. The result is the "test" package  is removed from the final build.  To notify others that you are doing this, the package name ends in with "test" (for example [zaptest](https://pkg.go.dev/go.uber.org/zap/zaptest) in uber's logging library)

Consider moving this the code to the "nitridingtest" package.

**Review by dlm — APPROVED** [review]
Major concerns that are easy to address.  Let me know if you need more info of want to discuss.

File: certificate/cert_internal_test.go [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+			errRegex: "x509: malformed certificate",
+		},
+		"malformed tbs certificate": {
+			derBytes: bytes.Repeat([]byte{0x30}, 50),
+			errRegex: "x509: malformed tbs certificate",
+		},
+		"malformed serial number ": {
+			derBytes: append([]byte{0x30, 0x30, 0x30, 0x20}, bytes.Repeat([]byte{0x30}, 46)...),
+			errRegex: "x509: malformed serial number",
+		},
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert_internal_test.go:124]: Thanks!

File: certificate/cert_internal_test.go [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+		require.NoError(t, err)
+		require.NotNil(t, pemBytes)
+		_, pemToDerErr := pemToDer(pemBytes)
+
+		ok, err := utils.ErrorMatchesPattern(
+			pemToDerErr,
+			"x509: malformed certificate",
+			"x509: malformed tbs certificate",
+			"x509: malformed serial number",
+		)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert_internal_test.go:162]: Thanks!

File: utils/limit_reader.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (mwittie) [reply utils/limit_reader.go:1]: Yes, I think we should have a discussion about this. I was gonna add an engineering item, but I see you already did that!

In this particular case, I think it doesn't matter, because the functionality of the code is fairly generic and doesn't expose any secret sauce functionality. Still a policy discussion would be useful.

File: utils/limit_reader.go (hunk 14 above)
Comment (mwittie) [reply utils/limit_reader.go]: Cool, I incorporated your updates. 

I also made a couple of changes you can see in https://github.com/brave/nitriding/commit/3b63971681345ddfd0266d5f50b4f9164152305f
One change I forgot to mention in the commit message is that I replace the deprecated `ioutil.ReadAll` to `io.ReadAll`. Since this is out of scope, I'm not touching ZPR code, but I'm happy to make a card to update https://bitbucket.org/blockymcchainerson/zpr/src/main/nub/input/reader.go with any of the changes, since you may not agree with all in the context of ZPR.

File: utils/testdata/exceeds-size.json (hunk 15 above)
Comment (mwittie) [reply utils/testdata/exceeds-size.json]: I updated to the newest version from ZPR. I think this addresses your comment, but I'm not sure. Please clarify if you think there's more to do here.

File: utils/limit_reader.go (hunk 14 above)
Comment (mwittie) [reply utils/limit_reader.go]: When linting I found another error with the `defer os.Close(file)` not handling an error. I read around an put together a potential solution. I don't fully love it and I'd like your feedback.
https://github.com/blocky/nitriding/commit/757c162a81da6c836b248bc8b60b8045608fe283
I could be convinced to go back to avoid the complexity.

Comment (mwittie) [reply utils/limit_reader.go]: OK, so this is all a moot point since there is not the `io.LimitReader` in go 1.19. I removed our implementation of `LimitedReader`

File: utils/limit_reader.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (mwittie) [reply utils/limit_reader.go]: Removed our `LimitedReader` with `io.LimitReader` from go 1.19

File: utils/testdata/exceeds-size.json (hunk 15 above)
Comment (mwittie) [reply utils/testdata/exceeds-size.json]: Removed our `LimitedReader` with `io.LimitReader` from go 1.19

File: validation/path_validator_test.go (hunk 16 above)
Comment (mwittie) [reply validation/path_validator_test.go]: I made it even better by making sure the lines are not too long :)

File: certificate/acme_cert_test.go [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+	"errors"
+	"testing"
+
+	"github.com/blocky/nitriding/certificate"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMakeACMECert_HappyPath(t *testing.T) {
+	selfSigCert, err := certificate.MakeSelfSigCert("", "")
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/acme_cert_test.go:13]: done

File: certificate/cert_internal_test.go [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+	assert.Equal(t, cert.DerBytes(), derBytes)
+}
+
+func TestPemToDer_CannotParsePEM(t *testing.T) {
+	_, err := pemToDer([]byte("not PEM data"))
+	assert.ErrorContains(t, err, "no PEM data found")
+}
+
+func FuzzPemToDer_CannotParsePEM(f *testing.F) {
+	tests := []struct {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert_internal_test.go]: When you Fuzz there will be more cases. But, we do need to seed the test corpus with something and make sure that input does something expected when running the test not in Fuzz mode.

File: certificate/cert_internal_test.go [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+
+	pemBytes, err := encodeToMemory(cert.DerBytes())
+	assert.NoError(t, err)
+	assert.NotNil(t, pemBytes)
+
+	VerifyCert(t, cert.DerBytes(), pemBytes)
+}
+
+func FuzzEncodeToMemory_NeverFails(f *testing.F) {
+	tests := []struct {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert_internal_test.go]: see my comment above :)

File: certificate/self_sig_cert.go [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+		return nil, err
+	}
+	err = pem.Encode(
+		pemFile,
+		&pem.Block{Type: "CERTIFICATE", Bytes: cert.derBytes},
+	)
+	if err != nil {
+		return nil, err
+	}
+	err = pemFile.Close()
```
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/self_sig_cert.go:162]: Not a perfect solution, but I moved `chmod` before `close`

File: utils/utils.go (hunk 18 above)
Comment (mwittie) [reply utils/utils.go:10]: done

File: certificate/cert_internal_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (dlm) [reply certificate/cert_internal_test.go]: I think you are missing my question.  You add cases to a fuzzer with `f.Add` not by using a pattern that is similar to a table testing.  If there is currently no other cases and you have run the fuzzer and saturated the test space the the point that we are satisfied with the result, do we really need 8 lines of setup code when 1 line will do?

File: certificate/self_sig_cert.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (dlm) [reply certificate/self_sig_cert.go:168]: Maybe I am confused but isn't line 167 setting the permission of the file?

[the remark this answers is not in the document]
Comment (dlm) [reply certificate/self_sig_cert.go:168]: Maybe I am confused... but isn't line 167 setting the file permission?

**Review by dlm — COMMENTED** [review]
somehow I started another review... closing review.

File: certificate/cert_internal_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert_internal_test.go]: Oh yeah, I did miss your question.

But, I tend to like to repeat the same syntactic setup for tests. Yes, it's more verbose, but because all of them have the same structure, it's easy to see what's different between them. You can also see that I'm wrapping the parameters in a structure. I don't actually need to do that either. I could do `f.Add([]byte("asdf")`, but I thought that having structures makes these tests more easily extendible, while keeping the syntax the same.

Thoughts?

File: certificate/self_sig_cert.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/self_sig_cert.go:168]: When you click the code link associated with the comment it takes you to the original file. Here's the code latest version of this function: https://github.com/blocky/nitriding/blob/feature/new-interfaces/certificate/self_sig_cert.go

[the remark this answers is not in the document]
Comment (dlm) [reply certificate/self_sig_cert.go:168]: oh... okay... now I see the new version!  Thanks for the info.

File: certificate/cert_internal_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (dlm) [reply certificate/cert_internal_test.go]: I have not used fuzzing in anything real enough to have a strong option. But I suspect that it will turn out similar to parameterized (aka table driven) tests.  That is, we don't parameterize every test just in case we come up with more than 1 case, but we parameterize when it becomes useful.

File: certificate/cert_internal_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert_internal_test.go]: fixed

File: certificate/cert_internal_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (mwittie) [reply certificate/cert_internal_test.go]: OK, that's a good comparison. I wouldn't follow that principle for table tests.

I updated to just use `f.Add([]byte("asdf")`, etc.

**Review by ianhecker — APPROVED** [review]

File: utils/limit_reader.go (hunk 14 above)
Comment (ianhecker) [reply utils/limit_reader.go]: looks good

File: validation/path_validator_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ianhecker) [reply validation/path_validator_test.go]: ok!

File: certificate/self_sig_cert_test.go (hunk 2 above)
Comment (ianhecker) [reply certificate/self_sig_cert_test.go]: ok

**Review by ianhecker — APPROVED** [review]
resolved comments
